src/serve/ouroboros/mod.rs

```diff
@@ -105,3 +83,140 @@ async fn handle_blockfetch(db: RollDB, mut peer: PeerServer) -> Result<(), Error
+            Point::Specific(slot, hash) => {
+                let hash: [u8; 32] = hash[0..32].try_into().unwrap();
+                let hash = Hash::<32>::from(hash);
+
+                if db.chain_contains(*slot, &hash).unwrap() {
+                    return Some(point.clone());
+                }
+            }
+        }
+    }
+
+    None
+}
+
+fn db_tip_to_protocol(tip: (u64, Hash<32>)) -> Tip {
+    // TODO: get block height from db
+    Tip(Point::Specific(tip.0, tip.1.to_vec()), 0)
+}
+
+#[instrument(skip_all)]
+async fn handle_chainsync(db: RollDB, mut protocol: chainsync::N2NServer) -> Result<(), Error> {
+    let mut intersect = db.crawl_chain_from(None);
+
+    loop {
+        debug!("waiting for request");
+
+        let req = protocol.recv_while_idle().await.map_err(Error::server)?;
+
+        debug!("new client request");
+
+        let tip = db
+            .find_tip()
+            .map_err(Error::server)?
+            .map(db_tip_to_protocol)
+            .unwrap_or(Tip(Point::Origin, 0));
+
+        debug!(?tip, "fetched tip from db");
+
+        match req {
+            Some(x) => match x {
+                chainsync::ClientRequest::Intersect(points) => {
+                    debug!(?points, "intersect request");
+
+                    if let Some(found) = find_valid_intersection(&db, &points) {
+                        intersect = match found {
+                            Point::Origin => db.crawl_chain_from(None),
+                            Point::Specific(x, _) => db.crawl_chain_from(Some(x)),
+                        };
+
+                        protocol
+                            .send_intersect_found(found, tip)
+                            .await
+                            .map_err(Error::server)?;
+                    } else {
+                        protocol
+                            .send_intersect_not_found(tip)
+                            .await
+                            .map_err(Error::server)?;
+                    }
+                }
+                chainsync::ClientRequest::RequestNext => {
+                    let next = intersect.next();
+
+                    if let Some(next) = next {
+                        let (_, hash) = next.map_err(Error::server)?;
+
+                        let block = db
+                            .get_block(hash)
+                            .map_err(Error::server)?
+                            .expect("block content not found");
+
+                        let block = MultiEraBlock::decode(&block).expect("invalid block cbor");
+
+                        let content = HeaderContent {
+                            variant: 1, // TODO
+                            byron_prefix: None,
+                            cbor: block.header().cbor().to_vec(),
+                        };
+
+                        protocol
+                            .send_roll_forward(content, tip)
+                            .await
+                            .map_err(Error::server)?;
```

Comment:
The first response to a RequestNext message the Haskell node makes after a successful intersect is a MsgRollback to the intersection point, so we probably want to copy that behaviour

```
    let known_points = vec![Point::Specific(
        43847831u64,
        hex::decode("15b9eeee849dd6386d3770b0745e0450190f7560e5159b1b3ab13b14b2684a45").unwrap(),
    )];

    let (point, _) = peer.chainsync().find_intersect(known_points).await.unwrap();

    info!("intersected point is {:?}", point);

    for _ in 0..3 {
        let next = peer.chainsync().request_next().await.unwrap();

        match next {
            chainsync::NextResponse::RollForward(h, _) => {
                info!("rolling forward, header size: {}", h.cbor.len())
            }
            chainsync::NextResponse::RollBackward(x, _) => info!("rollback to {:?}", x),
            chainsync::NextResponse::Await => info!("tip of chaing reached"),
        };
    }
```

```
2023-09-12T10:11:39.335692Z  INFO n2n_miniprotocols: intersected point is Some((43847831, 15b9eeee849dd6386d3770b0745e0450190f7560e5159b1b3ab13b14b2684a45))
2023-09-12T10:11:39.625239Z  INFO n2n_miniprotocols: rollback to (43847831, 15b9eeee849dd6386d3770b0745e0450190f7560e5159b1b3ab13b14b2684a45)
2023-09-12T10:11:39.958946Z  INFO n2n_miniprotocols: rolling forward, header size: 1006
2023-09-12T10:11:40.249326Z  INFO n2n_miniprotocols: rolling forward, header size: 1006
```